File: app/backend/services/interfaces/analytics_service.py
from typing import Protocol, runtime_checkable, Literal


@runtime_checkable
class IAnalyticsService(Protocol):
    async def get_overview(self, course_filter: list[str] | None = None) -> dict: ...

    async def get_activity(
        self,
        range_param: Literal["7d", "30d", "90d"],
        course_filter: list[str] | None = None,
    ) -> list[dict]: ...

    async def get_course_overview(self, course_id: str) -> dict: ...

    async def get_course_activity(self, course_id: str, range_param: Literal["7d", "30d", "90d"]) -> list[dict]: ...

    async def get_course_topics(self, course_id: str) -> list[dict]: ...

    async def get_course_sources(self, course_id: str) -> list[dict]: ...

File: app/frontend/src/services/analytics.service.ts
import * as analyticsApi from "@/api/analytics";
import type {
  OverviewRead,
  ActivityRead,
  CourseOverviewRead,
  CourseTopicsRead,
  CourseSourcesRead,
} from "@/types/analytics";

export const AnalyticsService = {
  async getOverview(): Promise<OverviewRead> {
    return analyticsApi.getOverview();
  },

  async getActivity(range: "7d" | "30d" | "90d"): Promise<ActivityRead> {
    return analyticsApi.getActivity(range);
  },

  async getCourseOverview(courseId: string): Promise<CourseOverviewRead> {
    return analyticsApi.getCourseOverview(courseId);
  },

  async getCourseActivity(courseId: string, range: "7d" | "30d" | "90d"): Promise<ActivityRead> {
    return analyticsApi.getCourseActivity(courseId, range);
  },

  async getCourseTopics(courseId: string): Promise<CourseTopicsRead> {
    return analyticsApi.getCourseTopics(courseId);
  },

  async getCourseSources(courseId: string): Promise<CourseSourcesRead> {
    return analyticsApi.getCourseSources(courseId);
  },
};
